### PR TITLE
Frontend Testing: Add tests for workspaces.ts reducer

### PR DESCRIPTION
**Pull request by jon-chua**
Part of #581 

- Added tests for workspaces.ts reducer
- Edited EVAL_TESTCASE_SUCCESS and EVAL_TESTCASE_FAILURE to prevent modifications to original array
- Edited BROWSE_HISTORY_REPL_UP and BROWSE_HISTORY_REPL_DOWN to avoid accessing negative indexes in arrays
- Removed EVAL_TESTCASE as it is not needed

### CODE REVIEW

File: src/reducers/__tests__/workspaces.ts [hunk 1]
```diff
@@ -0,0 +1,1169 @@
...
+  IPlaygroundWorkspace,
+  IWorkspaceManagerState,
+  maxBrowseIndex,
+  RunningOutput
+} from '../states';
+import { reducer } from '../workspaces';
+
+const assessmentWorkspace: WorkspaceLocation = WorkspaceLocations.assessment;
+const gradingWorkspace: WorkspaceLocation = WorkspaceLocations.grading;
+const playgroundWorkspace: WorkspaceLocation = WorkspaceLocations.playground;
```
[the remark this answers is not in the document]
Comment (jon-chua) [reply src/reducers/__tests__/workspaces.ts]: Done

File: src/reducers/__tests__/workspaces.ts [hunk 2]
```diff
@@ -0,0 +1,1169 @@
...
+        ...evalEditorDefaultState[location],
+        isRunning: true,
+        isDebugging: false
+      }
+    });
+  });
+});
+
+// Test data for EVAL_INTERPRETER_ERROR, EVAL_INTERPRETER_SUCCESS, EVAL_TESTCASE_SUCCESS and HANDLE_CONSOLE_OUTPUT
+const lastOutput1: RunningOutput[] = [
```
[the remark this answers is not in the document]
Comment (jon-chua) [reply src/reducers/__tests__/workspaces.ts]: Done

File: src/reducers/__tests__/workspaces.ts [hunk 3]
```diff
@@ -0,0 +1,1169 @@
...
+    type: 'running',
+    consoleLogs: ['console-log-test']
+  },
+  {
+    type: 'running',
+    consoleLogs: ['console-log-test-2']
+  }
+];
+
+const lastOutput2: InterpreterOutput[] = [
```
[the remark this answers is not in the document]
Comment (jon-chua) [reply src/reducers/__tests__/workspaces.ts]: Done

File: src/reducers/workspaces.ts [hunk 4]
```diff
@@ -339,6 +340,7 @@ export const reducer: Reducer<IWorkspaceManagerState> = (
         }
       };
     case EVAL_TESTCASE_SUCCESS:
+      const testcaseSuccessCopy = cloneDeep(state[location].editorTestcases);
```
Comment (geshuming) [inline src/reducers/workspaces.ts]: May I know why deep copy is used?

Comment (jon-chua) [reply src/reducers/workspaces.ts]: The current implementation seems to be modifying the original editorTestcases in the test file. After some discussion with @rrtheonlyone , we decided to use this implementation.

File: src/reducers/__tests__/workspaces.ts [hunk 5]
```diff
@@ -0,0 +1,1169 @@
...
+      [location]: {
+        ...evalEditorDefaultState[location],
+        isRunning: true,
+        isDebugging: false
+      }
+    });
+  });
+});
+
+// Test data for EVAL_INTERPRETER_ERROR, EVAL_INTERPRETER_SUCCESS, EVAL_TESTCASE_SUCCESS and HANDLE_CONSOLE_OUTPUT
```
[the remark this answers is not in the document]
Comment (jon-chua) [reply src/reducers/__tests__/workspaces.ts]: Attempted a fix.

File: src/reducers/__tests__/workspaces.ts [hunk 6]
```diff
@@ -0,0 +1,1169 @@
...
+      ...payload
+    },
+    playground: {
+      ...defaultWorkspaceManager.playground,
+      ...payload
+    }
+  };
+}
+
+test('BROWSE_REPL_HISTORY_DOWN works on non-null browseIndex and returns replValue to last value on no further records', () => {
```
[the remark this answers is not in the document]
Comment (jon-chua) [reply src/reducers/__tests__/workspaces.ts]: Attempted a fix.

File: src/reducers/__tests__/workspaces.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (rrtheonlyone) [reply src/reducers/__tests__/workspaces.ts]: Okay checked both of the above out!

1. I am guessing this is just a fail safe so we dont accidentally remove a task which is not `running`. We should be okay if we leave it.

2. This is done so as to not accidentally pass in `workspaceLocation` from the action payload to the object

So we are good here!

File: src/reducers/workspaces.ts (hunk 4 above)
Comment (jon-chua) [reply src/reducers/workspaces.ts]: Done

File: src/containers/PlaygroundContainer.ts [hunk 7]
```diff
@@ -52,7 +53,7 @@ const mapStateToProps: MapStateToProps<IStateProps, {}, IState> = state => ({
   externalLibraryName: state.workspaces.playground.playgroundExternal
 });
 
-const location: WorkspaceLocation = 'playground';
+const location: WorkspaceLocation = WorkspaceLocations.playground;
```
[the remark this answers is not in the document]
Comment (jon-chua) [reply src/containers/PlaygroundContainer.ts]: Updated.

File: src/reducers/states.ts [hunk 8]
```diff
@@ -104,6 +104,7 @@ export interface ISessionState {
 type ReplHistory = {
   browseIndex: null | number; // [0, 49] if browsing, else null
   records: string[];
+  originalReplValue: string;
```
[the remark this answers is not in the document]
Comment (jon-chua) [reply src/reducers/states.ts]: Renamed.

**Pull request merged**